src/org/zaproxy/zap/authentication/PostBasedAuthenticationMethodType.java

```diff
@@ -252,6 +260,43 @@ protected HttpMessage prepareRequestMessage(UsernamePasswordAuthenticationCreden
 			return requestMessage;
 		}
 
+		private void replaceAntiCsrfTokenAndCookieValueIfRequired(HttpMessage requestMessage) throws URIException, NullPointerException, HttpMalformedHeaderException {
+			if(extAntiCsrf == null) {
+				extAntiCsrf = Control.getSingleton().getExtensionLoader().getExtension(ExtensionAntiCSRF.class);
+			}
+			List<AntiCsrfToken> antiCsrfTokensOfFirstLoginMsg = null;
+			if(extAntiCsrf != null) {
+				antiCsrfTokensOfFirstLoginMsg = extAntiCsrf.getTokens(requestMessage);
+			}
+			AntiCsrfToken antiCsrfTokenOfFirstLoginMsg = null;
+			if(antiCsrfTokensOfFirstLoginMsg != null && antiCsrfTokensOfFirstLoginMsg.size() > 0) {
+				antiCsrfTokenOfFirstLoginMsg = antiCsrfTokensOfFirstLoginMsg.get(0);
+			}
+			
+			if (antiCsrfTokenOfFirstLoginMsg != null) {
+				this.requireAnitCsrfTokenWhenLogin = true;
+				URI loginURI = new URI(loginRequestURL, true, "UTF-8");
+				HttpMessage loginMsgToRegenerateAntiCsrfToken = new HttpMessage(loginURI);
+				loginMsgToRegenerateAntiCsrfToken.setRequestingUser(null);
+				try {
+					getHttpSender().sendAndReceive(loginMsgToRegenerateAntiCsrfToken);
+				} catch (IOException e) {
+					e.printStackTrace();
+				}
+
+				String regeneratedAntiCsrfTokenValue = extAntiCsrf.getTokenValue(loginMsgToRegenerateAntiCsrfToken, antiCsrfTokenOfFirstLoginMsg.getName());
+
+				String requestBody = requestMessage.getRequestBody().toString();
+				if (regeneratedAntiCsrfTokenValue != null) {
+					requestBody = requestBody.replace(
+							antiCsrfTokenOfFirstLoginMsg.getValue(), paramEncoder.apply(regeneratedAntiCsrfTokenValue));
+					requestMessage.getRequestBody().setBody(requestBody);
+				}
+				TreeSet<HtmlParameter> cookieParams = loginMsgToRegenerateAntiCsrfToken.getResponseHeader().getCookieParams();
+				requestMessage.setCookieParams(cookieParams);
```

Comment:
This should not be required if using the users' session, the cookies are set into the request when the message is sent.